Python 3 & better support for isolate.

stefano-maggiolo opened the pull request:
Sorry for two commits not fully related...

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cms-dev/cms/805)
<!-- Reviewable:end -->

stefano-maggiolo commented:
@andreyv @jonhte - closes #803

lerks commented:
I'll look into this shortly, hopefully by tomorrow.

FYI, the Travis error isn't your fault. I just triggered a new test on master and now it fails there too, in the same way. The error appears to be due to the fact that the version number of Postgres 10.0 (released a couple of weeks ago, but which as far as I can tell we are not using) cannot be correctly parsed by Psycopg2. Do you have any ideas?

stefano-maggiolo commented:
Mmm, [this](http://oshanebailey.jamaicandevelopers.com/2017/10/11/error-could-not-determine-postgresql-version-10-0/) seems relevant... might try to play with it tonight and see if that's the reason.

lerks commented:
Also the fact that we're telling Travis that it should setup an environment for Ruby may mess a few things up.

stefano-maggiolo commented:
https://github.com/psycopg/psycopg2/issues/489 is the actual source of truth - regex to detect version was incorrect, was fixed in 2.7.* but not backported to 2.6. Bumping from 2.6 to 2.7 makes the build green again.

I think our decision was to keep versions of python packages in sync with the latest LTS versions, but I think we can make an exception here. Opinions? In case I can add it to the PR.

lerks commented:
Nice find, that's the bug we're hitting! Apparently they didn't backport the fix to the 2.6 branch. We could either ask them to do that (and then depend on the next micro they release) or bump to 2.7. I'm OK with bumping (but please add a comment in requirements.txt explaining the reason). This could create issues to people running on the latest Ubuntu LTS using native packages (I know we recommend PIP and that it's sort-of their fault, but Ubuntu used to be our recommended environment and we pledged to support it out-of-the-box).

lerks commented on cms/grading/languages/python3_cpython.py:
```diff
@@ -0,0 +1,67 @@
+        basename = os.path.splitext(os.path.basename(source_filenames[0]))[0]
+        compilation_program = ";".join([
+            "import py_compile as m",
+            "m.compile(\"%s.py\", \"%s\", doraise=True)" % (
```
Here we're injecting unescaped strings into a shell command. It's true that they are coming from the admin, and it's true that they are run in a sandbox (are they?), though maybe we could escape them anyway? What do you think? Is the risk worth the effort?

andreyv replied:
I had thought about this too. One option is manual escaping (is there a built-in Python function for this?), another is to pass arguments on command line:
```
> python -c 'import sys;print(sys.argv[1])' test
test
```

lerks replied:
There sure is: `repr(unsafe_string)`.

lerks commented on cms/grading/Sandbox.py:
```diff
@@ -734,7 +734,17 @@ def translate_box_exitcode(self, _):
+        pass
+
     def delete(self):
+
```
Why this empty line before the docstring?

lerks commented on cms/grading/Sandbox.py:
```diff
@@ -842,14 +852,19 @@ def __init__(self, multithreaded, file_cacher, temp_dir=None):
+        # if the worker was interrupted in the middle of an execution.
+        cleanup_cmd = [self.box_exec] + (["--cg"] if self.cgroup else []) \
+            + ["--box-id=%d" % self.box_id] + ["--cleanup"]
+        subprocess.call(cleanup_cmd)
```
Is the return code interesting?

lerks replied:
And can't these three lines be replaced by a call to self.cleanup()?

lerks commented on cms/grading/Sandbox.py:
```diff
@@ -1314,17 +1329,24 @@ def translate_box_exitcode(self, exitcode):
         # Tell isolate to cleanup the sandbox.
         box_cmd = [self.box_exec] + (["--cg"] if self.cgroup else []) \
             + ["--box-id=%d" % self.box_id]
         subprocess.call(box_cmd + ["--cleanup"])
```
No need to have this argument be on its own.

stefano-maggiolo commented:
Bumped to 2.7 and added the comment.

---

Review status: 0 of 3 files reviewed at latest revision, 4 unresolved discussions, some commit checks broke.

---

*[cms/grading/Sandbox.py, line 747 at r1](https://reviewable.io:443/reviews/cms-dev/cms/805#-Kx5_UrPEwLdCJ2MlsEQ-r1-747:-Kxd_4Eh522pV7Av6Nid:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/3380066163b1ceee576542f01211ff46be81d952/cms/grading/Sandbox.py#L747)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Why this empty line before the docstring?
</blockquote></details>

Done.

---

*[cms/grading/Sandbox.py, line 860 at r1](https://reviewable.io:443/reviews/cms-dev/cms/805#-Kx5_UrPEwLdCJ2MlsEQ-r1-860:-Kxd_fUUACo4HxjpxT6t:b-paq7f8) ([raw file](https://github.com/cms-dev/cms/blob/3380066163b1ceee576542f01211ff46be81d952/cms/grading/Sandbox.py#L860)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

And can't these three lines be replaced by a call to self.cleanup()?
</blockquote></details>

Yes, replaced. Return code probably not - if there is an error we'll get an error in the subsequent init too, when really we cannot continue.

---

*[cms/grading/Sandbox.py, line 1342 at r1](https://reviewable.io:443/reviews/cms-dev/cms/805#-Kx5_UrPEwLdCJ2MlsEQ-r1-1342:-Kxda4-ERrClwoi4Z7BK:b-896fix) ([raw file](https://github.com/cms-dev/cms/blob/3380066163b1ceee576542f01211ff46be81d952/cms/grading/Sandbox.py#L1342)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

No need to have this argument be on its own.
</blockquote></details>

Done.

---

*[cms/grading/languages/python3_cpython.py, line 58 at r2](https://reviewable.io:443/reviews/cms-dev/cms/805#-Kx5_UrQdI0KeUhyltmo-r2-58:-KxdcloZmy7FUGTOhG43:b-s2rc7i) ([raw file](https://github.com/cms-dev/cms/blob/d5e68da68213299db1416ed2377e2f99bcb3c909/cms/grading/languages/python3_cpython.py#L58)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

There sure is: `repr(unsafe_string)`.
</blockquote></details>

Good point, thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/805#-:-KxdcrxFhLb0Pl_W1Oyu:b-4dsl05)*
<!-- Sent from Reviewable.io -->

lerks commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r3, 1 of 1 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/805#-:-Kxddj-FkwiG0pBfPZ8l:bnfp4nl)*
<!-- Sent from Reviewable.io -->
